File: backend/src/services/vocabulary-note.service.js
const Vocabulary = require("../models/vocabulary.model");
const vocabularyNoteRepository = require("../repositories/vocabulary-note.repository");

const MAX_LEN = 4000;

const getNote = async (userId, wordId) => {
  const v = await Vocabulary.findByPk(wordId);
  if (!v) {
    const e = new Error("vocabulary not found");
    e.statusCode = 404;
    throw e;
  }
  const row = await vocabularyNoteRepository.getNote(userId, wordId);
  return row || { note_text: null, updated_at: null };
};

const putNote = async (userId, wordId, noteText) => {
  const v = await Vocabulary.findByPk(wordId);
  if (!v) {
    const e = new Error("vocabulary not found");
    e.statusCode = 404;
    throw e;
  }
  const text = String(noteText ?? "").trim();
  if (text.length > MAX_LEN) {
    const e = new Error(`note_text max length ${MAX_LEN}`);
    e.statusCode = 400;
    throw e;
  }
  if (!text) {
    await vocabularyNoteRepository.deleteNote(userId, wordId);
    return { note_text: null, updated_at: null };
  }
  return vocabularyNoteRepository.upsertNote(userId, wordId, text);
};

const deleteNote = async (userId, wordId) => {
  const v = await Vocabulary.findByPk(wordId);
  if (!v) {
    const e = new Error("vocabulary not found");
    e.statusCode = 404;
    throw e;
  }
  await vocabularyNoteRepository.deleteNote(userId, wordId);
  return { deleted: true };
};

module.exports = {
  getNote,
  putNote,
  deleteNote,
};

File: mobile-app/lib/services/vocabulary_note_service.dart
import 'api_client.dart';

class VocabularyNoteService {
  VocabularyNoteService(this._api);

  final ApiClient _api;

  Future<String?> getNote(int wordId) async {
    try {
      final res = await _api.dio.get('/vocabulary/$wordId/note');
      final d = res.data['data'];
      if (d is Map) {
        final t = d['note_text'] as String?;
        if (t != null && t.trim().isNotEmpty) return t;
      }
    } catch (_) {}
    return null;
  }

  Future<void> saveNote(int wordId, String text) async {
    await _api.dio.put('/vocabulary/$wordId/note', data: {'note_text': text});
  }

  Future<void> deleteNote(int wordId) async {
    await _api.dio.delete('/vocabulary/$wordId/note');
  }
}
